File: src/bpm/vercmp.c
/* vercmp.c — package version comparison (alpm/rpmvercmp algorithm).
 *
 * Compares versions of the form [epoch:]pkgver[-pkgrel], e.g. "16.1.0-2".
 * Returns <0 if a<b, 0 if equal, >0 if a>b. This is what decides whether a
 * repo version is genuinely newer than what's installed (a plain strcmp would
 * call 1.10 < 1.9 and "upgrade" sideways/backwards).
 *
 * Algorithm: split each version into alternating runs of digits and letters,
 * separated by any non-alphanumeric. Compare run by run — numeric runs as
 * integers (leading zeros stripped), alpha runs lexically; a numeric run
 * outranks an alpha run; a longer numeric tail outranks a shorter one.
 */
#include "bpm.h"
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <ctype.h>

/* Compare one pkgver/pkgrel/epoch segment (no separators stripped specially). */
static int rpmvercmp(const char *a, const char *b) {
    if (!strcmp(a, b)) return 0;

    char *one = xstrdup(a), *two = xstrdup(b);
    char *p1 = one, *p2 = two;
    int ret = 0;

    while (*p1 || *p2) {
        while (*p1 && !isalnum((unsigned char)*p1)) p1++;
        while (*p2 && !isalnum((unsigned char)*p2)) p2++;

        /* If we ran out of one side, the one with more segments is newer. */
        if (!*p1 || !*p2) break;

        char *s1 = p1, *s2 = p2;
        int isnum;
        if (isdigit((unsigned char)*p1)) {
            while (isdigit((unsigned char)*p1)) p1++;
            while (isdigit((unsigned char)*p2)) p2++;
            isnum = 1;
        } else {
            while (isalpha((unsigned char)*p1)) p1++;
            while (isalpha((unsigned char)*p2)) p2++;
            isnum = 0;
        }

        char sep1 = *p1, sep2 = *p2;
        *p1 = '\0'; *p2 = '\0';

        /* One side numeric, the other alpha (or empty): numeric wins. */
        if (s1 == p1) { ret = isnum ? 1 : -1; break; }   /* p2 had the run */
        if (s2 == p2) { ret = isnum ? 1 : -1; break; }

        if (isnum) {
            while (*s1 == '0') s1++;
            while (*s2 == '0') s2++;
            size_t l1 = strlen(s1), l2 = strlen(s2);
            if (l1 != l2) { ret = l1 < l2 ? -1 : 1; *p1 = sep1; *p2 = sep2; break; }
        }
        int c = strcmp(s1, s2);
        if (c) { ret = c < 0 ? -1 : 1; *p1 = sep1; *p2 = sep2; break; }

        *p1 = sep1; *p2 = sep2;
    }

    if (ret == 0) {
        /* equal so far; whichever still has alphanumeric tail is newer */
        while (*p1 && !isalnum((unsigned char)*p1)) p1++;
        while (*p2 && !isalnum((unsigned char)*p2)) p2++;
        if (*p1 && !*p2) ret = 1;
        else if (!*p1 && *p2) ret = -1;
    }

    free(one); free(two);
    return ret;
}

/* Split "[epoch:]ver[-rel]" into epoch/ver/rel (into caller buffers). */
static void split_evr(const char *v, char *ep, char *ver, char *rel, size_t n) {
    ep[0] = '0'; ep[1] = '\0';
    const char *colon = strchr(v, ':');
    const char *start = v;
    if (colon) {
        size_t l = (size_t)(colon - v); if (l >= n) l = n - 1;
        memcpy(ep, v, l); ep[l] = '\0';
        start = colon + 1;
    }
    const char *dash = strrchr(start, '-');
    if (dash) {
        size_t l = (size_t)(dash - start); if (l >= n) l = n - 1;
        memcpy(ver, start, l); ver[l] = '\0';
        snprintf(rel, n, "%s", dash + 1);
    } else {
        snprintf(ver, n, "%s", start);
        rel[0] = '\0';
    }
}

int vercmp(const char *a, const char *b) {
    char ea[64], va[256], ra[64], eb[64], vb[256], rb[64];
    split_evr(a, ea, va, ra, sizeof va);
    split_evr(b, eb, vb, rb, sizeof vb);

    int c = rpmvercmp(ea, eb);          /* epoch dominates */
    if (c) return c;
    c = rpmvercmp(va, vb);              /* then pkgver */
    if (c) return c;
    if (*ra && *rb) return rpmvercmp(ra, rb);  /* then pkgrel, if both present */
    return 0;
}

File: src/bpm-rs/src/vercmp.rs
//! alpm/rpmvercmp version comparison — a direct port of the C `vercmp.c`.
//!
//! Compares `[epoch:]pkgver[-pkgrel]` (e.g. "16.1.0-2"). A plain string compare
//! would rank 1.10 below 1.9; this splits into alternating digit/letter runs and
//! compares run-by-run, numeric runs as integers.

use std::cmp::Ordering;

fn is_alnum(c: u8) -> bool {
    c.is_ascii_alphanumeric()
}

/// Compare one segment (epoch / ver / rel).
fn rpmvercmp(a: &str, b: &str) -> Ordering {
    if a == b {
        return Ordering::Equal;
    }
    let one = a.as_bytes();
    let two = b.as_bytes();
    let (mut i, mut j) = (0usize, 0usize);

    loop {
        while i < one.len() && !is_alnum(one[i]) {
            i += 1;
        }
        while j < two.len() && !is_alnum(two[j]) {
            j += 1;
        }
        if i >= one.len() || j >= two.len() {
            break;
        }

        let (s1, s2) = (i, j);
        let isnum = one[i].is_ascii_digit();
        if isnum {
            while i < one.len() && one[i].is_ascii_digit() {
                i += 1;
            }
            while j < two.len() && two[j].is_ascii_digit() {
                j += 1;
            }
        } else {
            while i < one.len() && one[i].is_ascii_alphabetic() {
                i += 1;
            }
            while j < two.len() && two[j].is_ascii_alphabetic() {
                j += 1;
            }
        }

        // One side started numeric where the other has an alpha run (or empty).
        if s1 == i {
            return if isnum { Ordering::Greater } else { Ordering::Less };
        }
        if s2 == j {
            return if isnum { Ordering::Greater } else { Ordering::Less };
        }

        let mut r1 = &one[s1..i];
        let mut r2 = &two[s2..j];
        if isnum {
            while r1.len() > 1 && r1[0] == b'0' {
                r1 = &r1[1..];
            }
            while r2.len() > 1 && r2[0] == b'0' {
                r2 = &r2[1..];
            }
            if r1.len() != r2.len() {
                return r1.len().cmp(&r2.len());
            }
        }
        match r1.cmp(r2) {
            Ordering::Equal => {}
            ord => return ord,
        }
    }

    // Whichever still has an alphanumeric run left is newer; a leftover alpha
    // run loses to "no run" (matches pacman's "1.0" > "1.0alpha").
    let rest1 = i < one.len();
    let rest2 = j < two.len();
    match (rest1, rest2) {
        (true, false) => {
            if one[i].is_ascii_alphabetic() {
                Ordering::Less
            } else {
                Ordering::Greater
            }
        }
        (false, true) => {
            if two[j].is_ascii_alphabetic() {
                Ordering::Greater
            } else {
                Ordering::Less
            }
        }
        _ => Ordering::Equal,
    }
}

fn split(v: &str) -> (&str, &str, &str) {
    // [epoch:]ver[-rel]
    let (epoch, rest) = match v.split_once(':') {
        Some((e, r)) => (e, r),
        None => ("0", v),
    };
    let (ver, rel) = match rest.rsplit_once('-') {
        Some((ve, re)) => (ve, re),
        None => (rest, "0"),
    };
    (epoch, ver, rel)
}

/// Full version compare over epoch, then ver, then rel.
pub fn vercmp(a: &str, b: &str) -> Ordering {
    let (ea, va, ra) = split(a);
    let (eb, vb, rb) = split(b);
    match rpmvercmp(ea, eb) {
        Ordering::Equal => {}
        o => return o,
    }
    match rpmvercmp(va, vb) {
        Ordering::Equal => {}
        o => return o,
    }
    rpmvercmp(ra, rb)
}

#[cfg(test)]
mod tests {
    use super::*;
    use std::cmp::Ordering::*;
    #[test]
    fn ordering() {
        assert_eq!(vercmp("1.10", "1.9"), Greater);
        assert_eq!(vercmp("16.1.0-2", "16.1.0-1"), Greater);
        assert_eq!(vercmp("1.0", "1.0"), Equal);
        assert_eq!(vercmp("2.0", "1.9"), Greater);
        assert_eq!(vercmp("1:1.0", "2.0"), Greater);
        assert_eq!(vercmp("1.0-1", "1.0-2"), Less);
    }
}
